### ISSUE TITLE
False positive in useless_asref: receiver needs deref

### ISSUE THREAD

**Issue by dtolnay** [rust-lang/rust-clippy#3480]
The following code is minimized from a socket I/O trait in Chrome OS:

```rust
trait Trait {
    fn as_ptr(&self);
}

impl<'a> Trait for &'a [u8] {
    fn as_ptr(&self) {
        self.as_ref().as_ptr();
    }
}

fn main() {
    Trait::as_ptr(&&[][..]);
    println!("success");
}
```

```console
warning: this call to `as_ref` does nothing
 --> src/main.rs:7:9
  |
7 |         self.as_ref().as_ptr();
  |         ^^^^^^^^^^^^^ help: try this: `self`
  |
  = note: #[warn(clippy::useless_asref)] on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_asref
```

The original code, `self.as_ref().as_ptr()`, performance an implicit deref from `&&[u8]` to `&[u8]` followed by `<[T] as AsRef<[T]>>::as_ref` followed by `<[T]>::as_ptr`.

Clippy's suggested code, `self.as_ptr()`, performs an unconditional recursive call to the trait method `<&[u8] as Trait>::as_ptr` which changes the meaning of the program.

```console
warning: function cannot return without recursing
 --> src/main.rs:6:5
  |
6 |     fn as_ptr(&self) {
  |     ^^^^^^^^^^^^^^^^ cannot return without recursing
7 |         self.as_ptr();
  |         ------------- recursive call site
  |
  = note: #[warn(unconditional_recursion)] on by default
  = help: a `loop` may express intention better if this is on purpose
```

---

*`clippy 0.0.212 (b2601be 2018-11-27)`*

**Comment by oli-obk** [comment]
We should not be reporting the lint in case there are adjustments on the argument to `as_ref`.
